node/src/components/consensus/era_supervisor.rs

```diff
@@ -463,6 +501,79 @@ where
     pub(crate) fn is_initialized(&self) -> bool {
         self.is_initialized
     }
+
+    fn handle_initialize_eras(
+        &mut self,
+        key_blocks: HashMap<EraId, BlockHeader>,
+        mut maybe_validators: Option<BTreeMap<PublicKey, U512>>,
+        timestamp: Timestamp,
+        genesis_start_time: Timestamp,
+    ) -> HashMap<EraId, ProtocolOutcomes<I, ClContext>> {
+        let mut result_map = HashMap::new();
+
+        for era_id in self.iter_past(self.current_era, self.bonded_eras * 2) {
+            let newly_slashed;
+            let validators;
+            let start_height;
+            let era_start_time;
+
+            if era_id.is_genesis() {
+                newly_slashed = vec![];
+                validators = maybe_validators
+                    .take()
+                    .expect("missing genesis era validators");
+                start_height = 0;
+                era_start_time = genesis_start_time;
+            } else {
+                let key_block = key_blocks.get(&era_id).expect("missing key block");
+                start_height = key_block.height() + 1;
+                era_start_time = key_block.timestamp();
+                if era_id == self.protocol_config.last_activation_point {
+                    newly_slashed = vec![];
+                    validators = maybe_validators
+                        .take()
+                        .expect("missing activation era validators");
+                } else {
+                    newly_slashed = key_block
+                        .era_end()
+                        .expect("key block must be a switch block")
+                        .equivocators
+                        .clone();
+                    validators = key_block
+                        .next_era_validator_weights()
+                        .expect("missing validators from key block")
+                        .clone();
+                }
```

Comment:
Can you add comments? What these 3 different cases represent?

Reply:
Done in https://github.com/CasperLabs/casper-node/pull/1023/commits/10ee3692c0bfc436cb953d2f6735070bd4f49f84.